src/file_upload.rs

```diff
@@ -0,0 +1,152 @@
+    };
+    Box::new(
+        field
+            .map_err(FileUploadErrorKind::MultipartError)
+            .fold(0i64, move |acc, bytes| {
+                let rt = file
+                    .write_all(bytes.as_ref())
+                    .map(|_| acc + bytes.len() as i64)
+                    .map_err(FileUploadErrorKind::IOError);
+                future::result(rt)
+            }),
+    )
+}
+
+/// Create new future to handle file as multipart data.
+fn handle_multipart(
+    item: multipart::MultipartItem<dev::Payload>,
+    mut file_path: PathBuf,
+    overwrite_files: bool,
+) -> Box<Stream<Item = i64, Error = FileUploadErrorKind>> {
+    match item {
+        multipart::MultipartItem::Field(field) => {
+            let filename = field
+                .headers()
+                .get(header::CONTENT_DISPOSITION)
+                .ok_or(FileUploadErrorKind::ParseError)
+                .and_then(|cd| {
+                    header::ContentDisposition::from_raw(cd)
+                        .map_err(|_| FileUploadErrorKind::ParseError)
+                })
+                .and_then(|content_disposition| {
+                    content_disposition
+                        .get_filename()
+                        .ok_or(FileUploadErrorKind::ParseError)
+                        .map(String::from)
+                });
+            let err = |e: FileUploadErrorKind| Box::new(future::err(e).into_stream());
+            match filename {
+                Ok(f) => {
+                    match fs::metadata(&file_path) {
+                        Ok(metadata) => {
+                            if !metadata.is_dir() || metadata.permissions().readonly() {
+                                return err(FileUploadErrorKind::InsufficientPermissions);
+                            }
+                        }
+                        Err(_) => {
+                            return err(FileUploadErrorKind::InsufficientPermissions);
+                        }
+                    }
+                    file_path = file_path.join(f);
+                    Box::new(save_file(field, file_path, overwrite_files).into_stream())
+                }
+                Err(e) => err(e),
+            }
+        }
+        multipart::MultipartItem::Nested(mp) => Box::new(
+            mp.map_err(FileUploadErrorKind::MultipartError)
+                .map(move |item| handle_multipart(item, file_path.clone(), overwrite_files))
+                .flatten(),
+        ),
+    }
+}
+
+/// Handle incoming request to upload file.
+/// Target file path is expected as path parameter in URI and is interpreted as relative from
+/// server root directory. Any path which will go outside of this directory is considered
+/// invalid.
+/// This method returns future.
+pub fn upload_file(req: &HttpRequest<crate::MiniserveConfig>) -> FutureResponse<HttpResponse> {
+    let app_root_dir = req.state().path.clone().canonicalize().unwrap();
+    let path = match Query::<QueryParameters>::extract(req) {
```

Comment:
Can't you extract this type safely?

Reply:
I don't think I can, but I could return some type of error if canonicalize() returns an error. My justification for this unwrap was, that server root path should be always valid and exactly same unwrap() happens in [main](https://github.com/svenstaro/miniserve/blob/master/src/main.rs#L116)